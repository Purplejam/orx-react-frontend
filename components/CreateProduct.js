import useForm from '../lib/useForm';
import Form from './styles/Form';
import gql from 'graphql-tag';
import {useMutation} from '@apollo/client';
import DisplayMessage from './ErrorMessage';
import {ALL_PRODUCTS_QUERY} from './Products';
import Router from 'next/router';

const CREATE_PRODUCT_MUTATION = gql`
	mutation CREATE_PRODUCT_MUTATION(
	 $name: String!,
	 $price: Int!,
	 $description: String!,
	 $image: Upload
	) {
		createProduct(data: {
			name: $name,
			price: $price,
			description: $description,
			photo: {create: {image: $image, altText: $name}}
		})
		{
			id
			name
			price
			description
		}
	}
`

export default function CreateProduct () {
	const {inputs, handleChange, resetInputs, clearInputs} = useForm({
		name: 'Enter a name',
		price: 0,
		description: 'Enter a description',
		image: ''
	});

	const [createProduct, {loading, error, data}] = useMutation(CREATE_PRODUCT_MUTATION, 
		{variables: inputs,
			refetchQueries: [{query: ALL_PRODUCTS_QUERY}]});

	
	return(
		<Form onSubmit={async(e) => {
			e.preventDefault();
			//submit the inputfield to the backend
			const res = await createProduct();
			clearInputs();
			Router.push({pathname: `/product/${res.data.createProduct.id}`});
			}}>

			<DisplayMessage error={error}/>
			<fieldset disabled={loading} aria-busy={loading}>
				 <label htmlFor="image">
						Image
						<input type="file" id="image" 
							name="image" onChange={handleChange}/>
					</label> 

				 <label htmlFor="name">
						Name
						<input type="text" id="name" 
							name="name" placeholder="Name" 
							value={inputs.name} onChange={handleChange}/>
					</label> 
		
					<label htmlFor="price">
						Price
						<input type="number" id="price" 
							name="price" placeholder="Price" 
							value={inputs.price} onChange={handleChange}/>
					</label>

					<label htmlFor="description">
						Description
						<input type="text" id="description" 
							name="description" placeholder="Description" 
							value={inputs.description} onChange={handleChange}/>
					</label>
		
					<button type="submit">+ Add Product</button>
			</fieldset>
		</Form>
		)
}

