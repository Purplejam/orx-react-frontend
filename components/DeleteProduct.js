import gql from 'graphql-tag';
import {useMutation} from '@apollo/client';

const DELETE_PRODUCT_MUTATION = gql`
	mutation DELETE_PRODUCT_MUTATION($id: ID!) {
		deleteProduct(id: $id) {
			id
			name
		}
	}
`

function update(cache, result) {
	cache.evict(cache.identify(result.data.deleteProduct));
}

export default function DeleteProduct({id, children}) {
	const [deleteProduct, {loading, error}] = useMutation(DELETE_PRODUCT_MUTATION, {
		variables: {
			id: id
		},
		update: update
	})

	return(
					<button disabled={loading} type="button" onClick={() => {
						if(confirm('Are you sure you want to delete rhis item?')) {
							//use mutation to delete
							deleteProduct().catch((error) => alert(error.message));
						}
						}}>
						{children}
					</button>
		)
}