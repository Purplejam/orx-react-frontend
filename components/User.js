import {useQuery} from '@apollo/client';
import gql from 'graphql-tag';

export const CURRENT_USER = gql`
	query CURRENT_USER {
	 authenticatedItem {
	    ... on User {
	      name
	      id
	      email
	      password_is_set
	      cart {
	      		id
	        quantity
	        product{
	          id
	          name
	          price
	          description
	          photo {
	            image {
	              publicUrlTransformed
	            }
	          }
	        }
	      }
	    }
	  }
	}
`

export default function useUser() {
const {data} = useQuery(CURRENT_USER);
return data?.authenticatedItem;
}

