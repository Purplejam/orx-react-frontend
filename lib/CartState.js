import {useState, createContext, useContext} from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;


//LocalProvider "wrapper"
function CartStateProvider({children}) {
	const [cartOpen, setCartOpen] = useState(false);

	function cartToggle() {
		setCartOpen(!cartOpen);
	}

	function openCart() {
		setCartOpen(true);
	}

	function closeCart() {
		setCartOpen(false);
	}

	return(
		<LocalStateProvider value={{cartOpen, cartToggle, openCart, closeCart}}>{children}</LocalStateProvider>
		)
}

//useCart hook
function useCart() {
	const all = useContext(LocalStateContext);
	return all;
}


export {CartStateProvider, useCart};