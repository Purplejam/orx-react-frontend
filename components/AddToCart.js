import {useMutation} from '@apollo/client';
import {CURRENT_USER} from './User';
import gql from 'graphql-tag';

const ADD_TO_CART = gql`
	mutation ADD_TO_CART($productId: ID!) {
		addToCart(productId: $productId) {
			id
		}
	}
	`

export default function AddToCart({id}) {
	const[addToCart, {loading}] = useMutation(ADD_TO_CART, {
		variables: {
			productId: id
		},
		refetchQueries: [
		{query: CURRENT_USER}
		]
	})
	return (
		<button disabled={loading} type="button" onClick={addToCart}>Add{loading && 'ing'} to cart ðŸ›’</button>
		)
}