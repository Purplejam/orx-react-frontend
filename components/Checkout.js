import styled from 'styled-components';
import {loadStripe} from '@stripe/stripe-js';
import {Elements, CardElement, useElements, useStripe} from '@stripe/react-stripe-js';
import SickButton from '../components/styles/SickButton.js';
import NProgress from 'nprogress';
import {useState} from 'react';
import gql from 'graphql-tag';
import {useMutation} from '@apollo/client';
import { useRouter } from 'next/router';
import {useCart} from '../lib/CartState.js';
import {CURRENT_USER} from '../components/User.js';

const CheckoutFormStyles = styled.form`
	box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
	border: 1px solid rgba(0,0,0,0.06);
	border-radius: 5px;
	padding: 1rem;
	display: grid;
	grid-gap: 1rem; 
`

const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		checkOut(token: $token) {
			id
			charge
			total
			items {
				id
				name
			}
		}
	}
	`

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);


export function CheckoutForm() {

const router = useRouter();
const stripe = useStripe();
const [error, setError] = useState();
const [loading, setLoading] = useState(false);
const elements = useElements();
const[checkOut, {error: graphQLError}] = useMutation(CREATE_ORDER_MUTATION, {
	refetchQueries: [{query: CURRENT_USER}]
})
const {closeCart} = useCart();
async function handleSubmit(e) {

	//1. Stop the form from submitting and turn the loader on
	e.preventDefault();
	console.log('We gotta do  some work!');
	
 //2. Start the page transition
 NProgress.start();
 setLoading(true);

	//3. Create the payment method via stripe (token backs here if successful)
	const {error, paymentMethod} = await stripe.createPaymentMethod({
		type: 'card',
		card: elements.getElement(CardElement)
	});
	console.log(paymentMethod);

	//4. Handle any errors from stripe 
	if(error) {
		NProgress.done();
		setLoading(false);
		setError(error);
		return;
	}
	//5. Send the token from step 3 to our keystone server via custom mutation
	const order = await checkOut({
		variables: {
			token: paymentMethod.id
		}
	})

	console.log(`Finished with the order`);
	console.log(order);
	//6. Change the page to view the order
	router.push({
		pathname: '/order',
		query: {id: order.data.checkOut.id}
	})
	//7. Closed the cart
	closeCart();

	//8. Turn the loader off
	NProgress.done();
 setLoading(false);
}
	return (
			<CheckoutFormStyles onSubmit={handleSubmit}>
				{error && <p style={{fontSize: 12}}>{error.message}</p>}
				{graphQLError && <p style={{fontSize: 12}}>{graphQLError.message}</p>}
				<CardElement/>
				<SickButton>Check Out Now</SickButton>
			</CheckoutFormStyles>
		)
}

export default function Checkout() {
return(
	<Elements stripe={stripeLib}>
		<CheckoutForm/>
	</Elements>
	)
}
