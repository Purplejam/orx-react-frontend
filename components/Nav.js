import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import useUser from './User';
import SignOut from './SignOut';
import {useCart} from '../lib/CartState';
import CartCount from './CartCount';
import calcTotalCartItems from '../lib/calcTotalCartItems';


export default function Nav() {
	const user = useUser();
    const {openCart} = useCart();

  return (
    <NavStyles>
    	<Link href='/products/1'>Products</Link>
    	{
    		user && (
		    	<>
		    		<Link href='/sell'>Sell</Link>
		    		<Link href='/orders'>Orders</Link>
		    		<Link href='/account'>Account</Link>
                    <SignOut/>
                    <button type="button" onClick={openCart}>My Cart<CartCount 
                        count={calcTotalCartItems(user?.cart)}/>
                    </button>
		    	</>
    			)
    		}

    	{
    		!user && (
		    	<>
		    		<Link href='/signin'>Sign In</Link>
		    	</>
    			)
    		}
    </NavStyles>
  );
}

