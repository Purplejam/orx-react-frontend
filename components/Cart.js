import styled from 'styled-components';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import useUser from './User';
import formatFunction from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import {useCart} from '../lib/CartState';
import CloseButton from './styles/CloseButton';
import RemoveFromCart from './RemoveFromCart';
import Checkout from './Checkout';


const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

export function CartItem({cartItem}) {
if(!cartItem.product) return null;
	return(
			<CartItemStyles>
				<img width="100" src={cartItem.product.photo.image.publicUrlTransformed} alt=""/>
				<div>
					<p>
						{formatFunction(cartItem.product.price * cartItem.quantity)} â€“
						<em>{cartItem.quantity} &times; {formatFunction(cartItem.product.price)} each</em>
					</p>
				</div>
				<RemoveFromCart id={cartItem.id}/>
			</CartItemStyles>
		)
}


export default function Cart() {
	const user = useUser();
	if (!user) return null;
	const {cartOpen, closeCart} = useCart();
	return(
		<CartStyles open={cartOpen}>
			<header>
				<Supreme>
					{user.name}'s cart
				</Supreme>
				<CloseButton type="button" onClick={closeCart}>&times;</CloseButton>
			</header>
			<ul>
				{user.cart.map((cartItem) => <CartItem key={cartItem.id} cartItem={cartItem}/>)}
			</ul>
			<footer>
				<p>{formatFunction(calcTotalPrice(user.cart))}</p>
				<Checkout/>
			</footer>
		</CartStyles>
		)
}