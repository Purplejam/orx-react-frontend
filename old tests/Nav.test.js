import {render, screen} from '@testing-library/react';
import {MockedProvider} from '@apollo/react-testing';
import Nav from '../components/Nav';
import {fakeUser, fakeCartItem} from '../lib/testUtils';
import {CURRENT_USER} from '../components/User';
import {CartStateProvider} from '../lib/CartState';


//make some micks for being logged out/in and logged in with cart items

const noSignedInMocks = [
	{
		request: {query: CURRENT_USER},
		result: {data: {authenticatedItem: null}},
	}
]

const signedInMocks = [
	{
		request: {query: CURRENT_USER},
		result: {data: {authenticatedItem: fakeUser()}},		
	}
]

const signedInMocksWithCart = [
	{
		request: {query: CURRENT_USER},
		result: {data: {authenticatedItem: fakeUser({
			cart: [fakeCartItem()]
		})}},		
	}
]

describe('<Nav/>', () => {
	it('Renders and minimal nav when signed out', () => {
		const {container, debug} = render(
				<CartStateProvider>
					<MockedProvider mocks={noSignedInMocks}>
						<Nav/>
					</MockedProvider>
				</CartStateProvider>
			);
		expect(container).toHaveTextContent('Sign In');
		const link = screen.getByText('Sign In');
		expect(link).toHaveAttribute('href', '/signin');

		const productsLink = screen.getByText('Products');
		expect(productsLink).toHaveAttribute('href', '/products/1');
	})

	it('Renders a full nav when signed in', async () => {
		const {container, debug} = render(
				<CartStateProvider>
					<MockedProvider mocks={signedInMocks}>
						<Nav/>
					</MockedProvider>
				</CartStateProvider>
			);
		await screen.findByText('Account');

		const sellLink = screen.getByText('Sell');
		expect(sellLink).toHaveAttribute('href', '/sell');
		expect(container).toMatchSnapshot();
		expect(container).toHaveTextContent('My Cart');
		expect(container).toHaveTextContent('Sign Out');
	})

	it('Renders the amount of items in the cart', async () => {
		const {container, debug} = render(
				<CartStateProvider>
					<MockedProvider mocks={signedInMocksWithCart}>
						<Nav/>
					</MockedProvider>
				</CartStateProvider>
			);
		await screen.findByText('Account');
		expect(container).toHaveTextContent('3');
		//...or
		expect(screen.getByText('3')).toBeInTheDocument();
	})
})