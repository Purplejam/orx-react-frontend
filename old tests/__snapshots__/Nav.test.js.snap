// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> Renders a full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 jofUXi"
  >
    <a
      href="/products/1"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <a>
      Sign Out
    </a>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-sc-xxvp4g-1 gwXiGr"
      >
        <div>
          <p
            class="CartCount__Dot-sc-xxvp4g-0 glPYDY count"
          >
            0
          </p>
        </div>
      </span>
    </button>
  </ul>
</div>
`;

exports[`<Nav/> renders a full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 jofUXi"
  >
    <a
      href="/products/1"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <a>
      Sign Out
    </a>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-sc-xxvp4g-1 gwXiGr"
      >
        <div>
          <p
            class="CartCount__Dot-sc-xxvp4g-0 glPYDY count"
          >
            0
          </p>
        </div>
      </span>
    </button>
  </ul>
</div>
`;
